### PR TITLE
Support for Search Runs API using "filter" string

### PR DESCRIPTION
**Pull request by mparkhe**
- proto update
- refactor search utils to work with both "anded_expression" and "filter" string
- unit tests for search utils

--- 

new tests run pass

```
tests/utils/test_search_utils.py::test_search_filter_basics PASSED
tests/utils/test_search_utils.py::test_anded_expression PASSED
tests/utils/test_search_utils.py::test_anded_expression_2 PASSED
tests/utils/test_search_utils.py::test_filter[metric.acc >= 0.94-parsed_filter0] PASSED
tests/utils/test_search_utils.py::test_filter[metrics.XYZ = 3-parsed_filter1] PASSED
tests/utils/test_search_utils.py::test_filter[params.model = 'LinearRegression'-parsed_filter2] PASSED
tests/utils/test_search_utils.py::test_filter[metrics.rmse < 1 and params.model_class = 'LR'-parsed_filter3] PASSED
tests/utils/test_search_utils.py::test_filter[`metric`.a >= 0.1-parsed_filter4] PASSED
tests/utils/test_search_utils.py::test_filter[`params`.model >= 'LR'-parsed_filter5] PASSED
tests/utils/test_search_utils.py::test_error_filter[metric.acc >= 0.94; metrics.rmse < 1-Multiple expression in filter] PASSED
tests/utils/test_search_utils.py::test_error_filter[m.acc >= 0.94-Invalid search expression type] PASSED
tests/utils/test_search_utils.py::test_error_filter[acc >= 0.94-Invalid filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[p.model >= 'LR'-Invalid search expression type] PASSED
tests/utils/test_search_utils.py::test_error_filter[model >= 'LR'-Invalid filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[metrics.A > 0.1 OR params.B = 'LR'-Invalid clause(s) in filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[metrics.A > 0.1 NAND params.B = 'LR'-Invalid clause(s) in filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[metrics.A > 0.1 AND (params.B = 'LR')-Invalid clause(s) in filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[`metrics.A > 0.1-Invalid clause(s) in filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[param`.A > 0.1-Invalid clause(s) in filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[`dummy.A > 0.1-Invalid clause(s) in filter string] PASSED
tests/utils/test_search_utils.py::test_error_filter[dummy`.A > 0.1-Invalid clause(s) in filter string] PASSED
```

### CODE REVIEW

**Comment by mparkhe** [comment]
@andrewmchen : you were right about need for explicit checks for `OR` and `(` ... etc type of checks.

**Comment by smurching** [comment]
Looking now!

File: mlflow/utils/search_utils.py [hunk 1]
```diff
@@ -1,45 +1,198 @@
...
+                "key": key,
+                "comparator": comparator,
+                "value": value
+            }
+        else:
+            raise MlflowException("Invalid search expression type '%s'" % key_type)
+
+    def _parse(self):
+        if self._filter_string:
+            parsed = sqlparse.parse(self._filter_string)
```
Comment (smurching) [inline mlflow/utils/search_utils.py]: The docs for sqlparse ([link](https://sqlparse.readthedocs.io/en/latest/api/#sqlparse.parse)) aren't super explicit about whether this can throw an exception - it might be safe to just assume this never throws, but to be defensive we could try-catch here & raise an MlflowException with an INTERNAL_ERROR & the sqlparse exception message (otherwise I think the client will see a generic Flask-generated error message like):

```
2019/03/01 18:18:21 ERROR mlflow.utils.rest_utils: API request to http://localhost:5000/api/2.0/preview/mlflow/runs/search failed with code 500 != 200, retrying up to 0 more times. API response body: <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Final//EN">
<title>500 Internal Server Error</title>
<h1>Internal Server Error</h1>
<p>The server encountered an internal error and was unable to complete your request.  Either the server is overloaded or there is an error in the application.</p>
```

File: mlflow/utils/search_utils.py [hunk 2]
```diff
@@ -1,45 +1,198 @@
...
-            raise Exception("Invalid metric type: '%s', expected float or double")
-        metric = next((m for m in run.data.metrics if m.key == key), None)
-        if metric is None:
+            # either "metric" or "parameter", since valid type
+            return entity_type
+
+    @classmethod
+    def _get_identifier(cls, identifier):
+        try:
+            entity_type, key = identifier.split(".")
```
Comment (smurching) [inline mlflow/utils/search_utils.py]: It looks like this throws a ValueError if there's more than one "." in the search filter because the result of split() will contain more than two elements (so e.g. use cases like searching on a param with a period in the value, like `metrics.modelClassName = 'sklearn.ensemble.RandomForestClassifier'`, will break). Is that intentional & if so should we document it in the protobuf so it shows up in public docs?

**Review by smurching — COMMENTED** [review]
Had a couple of comments but otherwise looks pretty good (thanks for writing some thorough tests!)

File: mlflow/utils/search_utils.py (hunk 1 above)
Comment (mparkhe) [reply mlflow/utils/search_utils.py]: Added the extra check--but `sqlparse` is super robust and can parse any string. None string objects are dealt with due to type checking in protos

File: mlflow/utils/search_utils.py (hunk 2 above)
Comment (mparkhe) [reply mlflow/utils/search_utils.py]: No. This is applicable only for the left hand side if the expression. So if the filter were defined like

```
metric.crazy.name <= 0.01
```

And will error out. Will clean up message and add a test case.

**Review by smurching — APPROVED** [review]
Thanks for the updates! LGTM, let's just make sure we have a follow-up tracked somewhere to handle the case of searching over a param/metric with "." in its name
